x/twap/client/cli/query.go

```diff
@@ -0,0 +1,138 @@
+func GetQueryTwapCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "twap [poolid] [base denom] [start time] [end time]",
+		Short: "Query twap",
+		Long: strings.TrimSpace(
+			fmt.Sprintf(`Query twap for pull. Start time must be unix time. End time can be unix time or duration.
+
+Example:
+$ %s q twap 1 uosmo 1667088000 24h
+$ %s q twap 1 uosmo 1667088000 1667174400
+`,
+				version.AppName, version.AppName,
+			),
+		),
+		Args: cobra.ExactArgs(4),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// boilerplate parse fields
+			// <UINT PARSE>
+			poolId, err := parseUint(args[0], "poolId")
+			if err != nil {
+				return err
+			}
+
+			// <DENOM PARSE>
+			baseDenom := strings.TrimSpace(args[1])
+
+			// <UNIX TIME PARSE>
+			startTime, err := parseUnixTime(args[2], "start time")
+			if err != nil {
+				return err
+			}
+
+			// END TIME PARSE: ONEOF {<UNIX TIME PARSE>, <DURATION>}
+			endTime, err := parseUnixTime(args[3], "end time")
+			if err != nil {
+				// TODO if we don't use protoreflect:
+				// make better error combiner, rather than just returning last error
+				duration, err := time.ParseDuration(args[3])
+				if err != nil {
+					return err
+				}
+				endTime = startTime.Add(duration)
+			}
+
+			clientCtx, err := client.GetClientQueryContext(cmd)
+			if err != nil {
+				return err
+			}
+			queryClient := queryproto.NewQueryClient(clientCtx)
+			gammClient := gammtypes.NewQueryClient(clientCtx)
+			liquidity, err := gammClient.TotalPoolLiquidity(cmd.Context(), &gammtypes.QueryTotalPoolLiquidityRequest{PoolId: poolId})
+			if err != nil {
+				return err
+			}
+			if len(liquidity.Liquidity) != 2 {
+				return fmt.Errorf("pool %d has %d assets of liquidity, CLI support only exists for 2 assets right now.", poolId, len(liquidity.Liquidity))
+			}
+			quoteDenom := ""
+			if liquidity.Liquidity[0].Denom == baseDenom {
+				quoteDenom = liquidity.Liquidity[1].Denom
+			} else if liquidity.Liquidity[1].Denom == baseDenom {
+				quoteDenom = liquidity.Liquidity[0].Denom
+			} else {
+				return fmt.Errorf("pool %d doesn't have provided baseDenom %s, has %s and %s",
+					poolId, baseDenom, liquidity.Liquidity[0], liquidity.Liquidity[1])
+			}
+
+			res, err := queryClient.ArithmeticTwap(cmd.Context(), &queryproto.ArithmeticTwapRequest{
+				PoolId:     poolId,
+				BaseAsset:  baseDenom,
+				QuoteAsset: quoteDenom,
+				StartTime:  startTime,
+				EndTime:    &endTime,
+			})
+			if err != nil {
+				return err
+			}
+
+			return clientCtx.PrintProto(res)
+		},
+	}
+
+	flags.AddQueryFlagsToCmd(cmd)
+
+	return cmd
+}
+
+func parseUint(arg string, fieldName string) (uint64, error) {
+	v, err := strconv.ParseUint(arg, 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("could not parse %s as uint for field %s: %w", arg, fieldName, err)
+	}
+	return v, nil
+}
+
+func parseUnixTime(arg string, fieldName string) (time.Time, error) {
+	timeUnix, err := strconv.ParseInt(arg, 10, 64)
+	if err != nil {
```

Comment:
Maybe we should bring these into osmoutils at some point since we're doing the same thing in all cli files 🤔

Reply:
yeah, I think we either make querygen create this, or we use the protoreflect automation tooling from SDK to remove all this code in the first place.